Fragments: Separate out item bans

hakusaro opened the pull request:
This is the work in progress pull request for splitting item bans out into its own subsystem. Fixes #1563.

Sources of remaining references:
- [x] Chest handling in `GetDataHandlers`
- [ ] Bouncer
- [x] Misc checks in `TSPlayer`
- [ ] TShock init contains references to `ItemManager` that shouldn't really be there if we've done this properly
- [ ] `Utils` can reference `ItemManager` as well
- [ ] Commands maybe?

hakusaro committed:
Remove commented out warning disable

hakusaro committed:
Add initial ItemBans segregation infrastructure

hakusaro committed:
Add shell for initial OnSecondUpdate stuff

hakusaro committed:
Add comments yo

hakusaro committed:
Remove duplicated logic

hakusaro committed:
Split out more item ban code

This part of the fragments work is primarily aimed at reducing the
complexity of OnSecondUpdate in TShock and moving that check out into
the ItemBans subsytem.

Of major note in this is the removal of "check", which was a string
variable that tracked state and replacement of many of the item ban
activities with sane private methods that are at least somewhat
sensible. Obviously there's a lot to be desired in this system and I'm
really going for a run here by trying to continue a branch from so long
ago that I barely even remember the whole point of existence.

Still to do: GetDataHandlers related item ban code needs to be moved
into its own hook in the ItemBan system. Finally, there is a downside to
some of this: we're basically iterating over players again and again if
we keep this pattern up, which is kinda lame for complexity purposes.

hakusaro committed:
alt j: comment changes

hakusaro committed:
Move item ban check out of main playerupdate check

Separates out item ban logic from the rest of GetDataHandlers so that
item bans is more isolated in terms of what fragments is asking for.

hakusaro committed:
alt-j: convert indentation to tabs

hakusaro committed:
alt-j: fix botching source code

hakusaro committed:
Move item ban related chest checks out of gdh

hakusaro committed:
Remove chest item change detection from item bans

It doesn't do anything. If a user removes an item from a chest, it
bypasses this check. If a user adds an item to a chest, the server seems
to persist the change anyway, even if the event is handled. That's a bug
for sure, but fundamentally, it's not the item ban system's fault.

hakusaro committed:
Revert "Remove chest item change detection from item bans"

This reverts commit 758541ac5c4d4096df2db05ba2a398968113e1e4.

hakusaro committed:
Fix logic issues related to item ban handling

Re-implements chest item handling and correctly handles events and
returns after setting handled event state.

hakusaro committed:
Remove TSPlayer.HasProjectilePermission

In infinite wisdom, it turns out this is not a good method for TSPlayer
to have. It just checks the states of things as per what the item ban
system says is banned and then creates implicit relationships to the
projectile ban system.

Doing this effectively knocks down another external reference to the
item ban system outside of the context of the implementation for the
system itself and its related hooks.

This commit also adds context around what the heck is going on with some
of our more interesting checks as per discussions in Telegram with @Ijwu
and @QuiCM.

hakusaro committed:
Update changelog

hakusaro commented:
This is ready for review even though it's not fully done. I'd rather get it in gen-dev before we start contaminating the codebase with 1.4 changes. I think this makes everything better than it currently is right now and we can do minor patches later to improve quality.

mistzzt commented on TShockAPI/Commands.cs:
```diff
@@ -3274,7 +3274,31 @@ private static void ItemBan(CommandArgs args)
+							// command layer instead of inferring the state of projectile
+							// bans based on the state of the item ban system.
+
+							if (nameForStorage == "Dirt Rod")
```
Is this string comparison necessary? we can do `items[0].type == ItemID.blahblah` though

mistzzt commented on TShockAPI/Bouncer.cs:
```diff
@@ -803,8 +803,28 @@ internal void OnNewProjectile(object sender, GetDataHandlers.NewProjectileEventA
+			// without PvP enabled and belong to enemy mobs, so they shouldn't be
+			// possible for players to create. (Source: Ijwu, QuiCM)
+			Projectile tempProjectile = new Projectile();
+			tempProjectile.SetDefaults(type);
```
I went through the procedure but didn't see anywhere `tempProjectile` is referenced -- we only check `type` right?

hakusaro replied:
Holy ravioli, you're right. In `master` (pre-bouncer), this is what we had:
![Screenshot from 2020-05-16 16-05-41](https://user-images.githubusercontent.com/532647/82131915-2bd94c00-978f-11ea-8d83-22168bf59838.png)
I'll remove that.

mistzzt reviewed (approved):
lgtm so far

hakusaro committed:
Remove useless ref to Projectile.SetDefaults

hakusaro committed:
Change item ban to ban based on ID not strings

I think I was so confused as to why we were passing strings everywhere
that I just felt inclined to continue the trend in previous commits.

mistzzt commented:
Is this ready for getting into the codebase? I'm not sure so I left it without merging here.

hakusaro commented:
@mistzzt if you're good with it, you can squash and merge it so that it'll just be 1 commit. I'm personally happy with it going in now because it's all tested and working, and I really don't want to let it get left behind with the 1.4 patching
